### PR TITLE
Support path objects (PEP 519) in mmio functions

### PR DESCRIPTION
**Pull request by mruffalo**
#### Reference issue
Extends #6855 and #7649 with more I/O functions that I hadn't noticed/used until recently.

#### What does this implement/fix?
Allows using `pathlib.Path` objects (and more) in `scipy.io.mmio` functions. Added test coverage for passing `pathlib.Path` objects to `scipy.io.mmread` and `scipy.io.mmwrite`.

#### Additional information
I noticed in `setup.py` that SciPy now requires Python 3.6 or newer. That simplifies the implementation of this functionality a little bit, since we can now assume the existence of the `os.fspath` function, which converts any path-like object to a `str`. The approach here is "try to convert to `str`, if that fails it must be an open file-like object". (This approach can be used elsewhere, but that seems like it should be a separate PR.)

### CODE REVIEW

**Comment by mruffalo** [comment]
Ping -- is it okay to require Python 3.6 or newer? @andyfaff said in #11436 that "Scipy is still supporting Python 3.5", but `setup.py` on the current `master` branch disagrees: https://github.com/scipy/scipy/blob/52757d273bab9071a833bdef564f258cc0346d04/setup.py#L30-L31

**Comment by rgommers** [comment]
Yes that's okay, we dropped Python 3.5 in master

**Comment by mruffalo** [comment]
Is anything blocking this from being merged? I just rebased to `master` to trigger a new Travis CI build, which now passes (possibly after fixes for #11453).

I had just absentmindedly tried to pass a `pathlib.Path` to `scipy.io.mmread`, which failed and reminded me that this PR is still outstanding.

**Review by tylerjereddy — APPROVED** [review]
Looks sensible at first glance & no old tests were modified beyond cleanups for Python < 3.6. Probably no major blocker beyond expert reviewer bandwidth.

I'll approve as non-expert and add the `1.5.0` milestone to reduce the likelihood that we forget it in the next major release.

File: scipy/io/tests/test_paths.py [hunk 1]
```diff
@@ -70,6 +62,22 @@ def test_hb_write(self):
...
 
+    def test_mmio_read(self):
+        # Save data with string path, load with pathlib.Path
+        with tempdir() as temp_dir:
+            data = scipy.sparse.csr_matrix(scipy.sparse.eye(3))
+            path = Path(temp_dir) / 'data.mtx'
+            scipy.io.mmwrite(str(path), data)
+
+            data_new = scipy.io.mmread(path)
+            assert_((data_new != data).nnz == 0)
```
[the remark this answers is not in the document]
Comment (mruffalo) [reply scipy/io/tests/test_paths.py:73]: Thanks! I'll be happy to push a new commit that uses plain `assert`. Will do momentarily.

**Comment by perimosocordiae** [comment]
Looks good to me as well, so I pushed the button. Thanks for following up on this, @mruffalo !

**Pull request merged**